### PR TITLE
feat: lifecycle query runner class

### PR DESCRIPTION
**Pull request by mariusandra**
## Problem

Creating a structure for HogQL insights to follow.

## Changes

Creates an abstract QueryRunner, and its implementation in the shape of a LifecycleQueryRunner.

## How did you test this code?

Old tests kept working. Added back Robbie's events query tests I removed during a previous cleanup.

### CODE REVIEW

File: posthog/hogql_queries/lifecycle_query_runner.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        return LifecycleQueryResponse(result=res, timings=response.timings)
+
+    @cached_property
+    def query_date_range(self):
+        return QueryDateRange(
+            date_range=self.query.dateRange, team=self.team, interval=self.query.interval, now=datetime.now()
+        )
+
+    @cached_property
+    def event_filter(self) -> ast.Expr:
```
Comment (thmsobrmlr) [inline posthog/hogql_queries/lifecycle_query_runner.py:135]: Some of the methods are getting quite long and hard to read. I'd love if we could split them out into separate functions e.g. `_append_date_range`, `_append_properties` etc. at some point.

Timings could then also be done as decorators to help readability.

File: posthog/hogql_queries/lifecycle_query_runner.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    def run(self) -> LifecycleQueryResponse:
+        response = execute_hogql_query(
+            query_type="LifecycleQuery",
+            query=self.to_query(),
+            team=self.team,
+            timings=self.timings,
+        )
+
+        # TODO: can we move the data conversion part into the query as well? It would make it easier to swap
+        # e.g. the LifecycleQuery with HogQLQuery, while keeping the chart logic the same.
```
Comment (thmsobrmlr) [inline posthog/hogql_queries/lifecycle_query_runner.py:96]: Handling the response mangling on the app layer seems the right place to me 🤔

**Review by thmsobrmlr — APPROVED** [review]
Nice!

File: posthog/hogql_queries/lifecycle_query_runner.py (hunk 1 above)
Comment (mariusandra) [reply posthog/hogql_queries/lifecycle_query_runner.py:135]: Agreed. For now, I wanted to keep everything tight and packed. When we add the trends query, it's likely going to be too big for one file, so we're going to have to refactor things further... including likely moving event filters to their own files, and splitting them further.

File: posthog/hogql_queries/lifecycle_query_runner.py (hunk 2 above)
Comment (mariusandra) [reply posthog/hogql_queries/lifecycle_query_runner.py:96]: It does... but if we want to be able to move towards a flexible system where we can swap out `LifecycleQuery` with `HogQLQuery` at will, we'll need a layer that adds these custom transformations on top.

The options we have are
a) Avoid the problem by having everything in SQL (might not be possible/practical... and how will `InsightVizNode` know what to render if we just give a query?) 
b) `{ kind: 'LifecycleQuery', source: { ...HogQLQuery } }`
c) `{ kind: 'InsightDataSourceQuery', type: 'Lifecycle', source: { ...HogQLQuery }  }` as we need to tell the `InsightVizNode` _somehow_ what to render
d) ...? 🤔

**Pull request merged**